File: 16-3sum-closest/16-3sum-closest.java

// [-4, -1, 1, 2, 3, 4] target = 6
//       i        l   r  curSum = 6, curDiff = 1, difference = 1

//
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int difference = Integer.MAX_VALUE;
        int left;
        int right;
        int curSum;
        int curDiff;
        Arrays.sort(nums);
        
        
        
        for (int i = 0; i < nums.length-2; i++) {
            left = i+1;
            right = nums.length-1;
            
            
            while(left < right) {
                curSum = nums[i] + nums[left] + nums[right];
                curDiff = target - curSum;
                if(Math.abs(curDiff) < Math.abs(difference)) {
                    difference = curDiff;
                }
                
                
                if(curSum == target) {
                    return curSum;
                } else if(curSum > target) {
                    right--;
                } else {
                    left++;
                }
            }
        }
        
        
        return target - difference;
    }
}

File: 16-3sum-closest/16-3sum-closest.py
'''
 -4 -1 1 2  -> target = 1
     i j k 
  
cur_sum = 2
closest = -1


 if cur_sum < target
    j += 1
 elif cur_sum > target
    k -= 1
'''
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        closest_sum = float(inf)
        nums.sort()
        for i in range(len(nums) - 2):
            j = i + 1
            k = len(nums) - 1
            
            while j < k:
                
                cur_sum = nums[i] + nums[j] + nums[k]

                if abs(target - cur_sum) < abs(target - closest_sum):
                    closest_sum = cur_sum
                
                if cur_sum < target:
                    j += 1
                elif cur_sum > target:
                    k -= 1
                else:
                    return nums[i] + nums[j] + nums[k]
                
        return closest_sum